rotkehlchen/chain/ethereum/modules/uniswap/v3/decoder.py

```diff
@@ -57,93 +69,245 @@ def _maybe_decode_v3_swap(  # pylint: disable=no-self-use
+        """
+        if tx_log.topics[0] != SWAP_SIGNATURE:
+            return
+
+        # Uniswap V3 represents the delta of tokens in the pool with a signed integer
+        # for each token. In the transaction we have the difference of tokens in the pool
+        # for the token0 [0:32] and the token1 [32:64]. If that difference is negative it
+        # means that the tokens are leaving the pool (the user receives them) and if it is
+        # positive they get into the pool (the user sends them to the pool)
+        delta_token_0 = hex_or_bytes_to_int(tx_log.data[0:32], signed=True)
+        delta_token_1 = hex_or_bytes_to_int(tx_log.data[32:64], signed=True)
+        if delta_token_0 > 0:
+            amount_sent, amount_received = delta_token_0, -delta_token_1
+        else:
+            amount_sent, amount_received = delta_token_1, -delta_token_0
+
+        # Uniswap V3 pools are used with complex routers/aggregators and there can be
+        # multiple spent and multiple receipt events that are hard to decode by looking only
+        # at a single swap event. Because of that here we decode only basic info, leaving the rest
+        # of the work to the router/aggreagator-specific decoding methods.
+        decode_basic_uniswap_info(
+            amount_sent=amount_sent,
+            amount_received=amount_received,
+            decoded_events=decoded_events,
+            counterparty=CPT_UNISWAP_V3,
+            notify_user=self.notify_user,
+        )
+    # --- Routers methods ---
+
+    def _decode_eth_to_token_swap(
+            self,
+            decoded_events: list[HistoryBaseEntry],
+    ) -> Optional[SwapData]:
+        """
+        Decode a swap of ETH to a token. Such swap consists of 3 events:
+        1. Sending ETH to the router.
+        2. If there is a rfund, receiving ETH from the router.
+        3. Receiving tokens from the router.
+        """
+        send_eth_event = decoded_events[1]
+        refund_eth_event = decoded_events[2] if decoded_events[2].asset == A_ETH else None
+        receive_token_event = decoded_events[2] if refund_eth_event is None else decoded_events[3]
+        from_amount = send_eth_event.balance.amount
+
+        if refund_eth_event is not None:
+            from_amount -= refund_eth_event.balance.amount
+
+        return SwapData(
+            from_asset=A_ETH,
+            from_amount=from_amount,
+            to_asset=receive_token_event.asset,
+            to_amount=receive_token_event.balance.amount,
+        )
+
+    def _decode_token_to_eth_swap(
+            self,
+            decoded_events: list[HistoryBaseEntry],
+    ) -> Optional[SwapData]:
+        from_asset: Optional[Asset] = None
+        from_amount = ZERO
+
+        # In this loop we find all the events that spend the token and sum them up.
+        # `0` event is the gas event, `1` event is the ETH receipt event, so we start with `2`.
+        for event in decoded_events[2:]:
```

Comment:
why not just iterate over all the events? If we change anything we might affect the logic without noticing

Reply:
By iterating only specific events that we need I tried to make the logic more readabale and understandable.

Reply:
Leaving this comment open. For me is not more redeable and is not easier to understand since you need a comment to explain it